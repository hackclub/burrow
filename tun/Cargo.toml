[package]
name = "tun"
version = "0.1.0"
edition = "2021"

[dependencies]
libc = "0.2"
fehler = "1.0"
nix = { version = "0.26", features = ["ioctl"] }
socket2 = "0.4"
tokio = { version = "1.28", features = [] }
byteorder = "1.4"
tracing = "0.1"
log = "0.4"
serde = { version = "1", features = ["derive"], optional = true }

futures = { version = "0.3.28", optional = true }

[features]
serde = ["dep:serde"]
tokio = ["tokio/net", "dep:futures"]

[target.'cfg(feature = "tokio")'.dev-dependencies]
tokio = { features = ["rt", "macros"] }

[target.'cfg(windows)'.dependencies]
lazy_static = "1.4"
libloading = "0.7"
tempfile = "3.5"
widestring = "1.0"
windows = { version = "0.48", features = [
    "Win32_Foundation",
    "Win32_NetworkManagement_IpHelper",
] }

[target.'cfg(windows)'.build-dependencies]
anyhow = "1.0"
bindgen = "0.65"
reqwest = { version = "0.11", features = ["native-tls"] }
ssri = { version = "9.0", default-features = false }
tokio = { version = "1.28", features = ["rt"] }
zip = { version = "0.6", features = ["deflate"] }
