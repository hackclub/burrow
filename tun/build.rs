#[cfg(not(windows))]
fn main() {
    println!("cargo:rerun-if-changed=build.rs");
}

#[cfg(windows)]
#[tokio::main(flavor = "current_thread")]
async fn main() -> anyhow::Result<()> {
    let out_dir = std::env::var("OUT_DIR")?;
    let path = std::path::PathBuf::from(out_dir);
    generate(&path).await?;

    println!("cargo:rerun-if-changed=build.rs");
    Ok(())
}

#[cfg(windows)]
async fn generate(out_dir: &std::path::Path) -> anyhow::Result<()> {
    use std::{fs::File, io::Write};

    use anyhow::Context;

    let bindings_path = out_dir.join("wintun.rs");
    let binary_path = out_dir.join("wintun.dll");
    println!("cargo:rerun-if-changed={}", bindings_path.to_str().unwrap());
    println!("cargo:rerun-if-changed={}", binary_path.to_str().unwrap());

    if let (Ok(..), Ok(..)) = (File::open(&bindings_path), File::open(&binary_path)) {
        return Ok(())
    };

    let archive = download(out_dir)
        .await
        .context("Failed to download wintun")?;

    let (bindings, binary) = parse(archive).context("Failed to parse wintun archive")?;

    bindings
        .write_to_file(bindings_path)
        .context("Failed to write bindings")?;

    let mut file = std::fs::OpenOptions::new()
        .write(true)
        .create(true)
        .open(binary_path)
        .context("Failed to write binary")?;
    file.write_all(&binary)?;

    Ok(())
}

#[cfg(windows)]
async fn download(directory: &std::path::Path) -> anyhow::Result<std::fs::File> {
    use std::{io::Write, str::FromStr};

    let path = directory.join(WINTUN_FILENAME);
    let mut file = match std::fs::OpenOptions::new().read(true).open(&path) {
        Ok(existing) => return Ok(existing),
        Err(_e) => std::fs::OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(path)?,
    };

    let mut url = reqwest::Url::from_str("https://www.wintun.net/builds")?;
    url.path_segments_mut().unwrap().push(WINTUN_FILENAME);

    let body = reqwest::get(url).await?.bytes().await?;

    ssri::IntegrityChecker::new(WINTUN_INTEGRITY.parse()?)
        .chain(&body)
        .result()?;

    file.set_len(0)?;
    file.write_all(&body)?;

    Ok(file)
}

#[cfg(windows)]
fn parse(file: std::fs::File) -> anyhow::Result<(bindgen::Bindings, Vec<u8>)> {
    use std::io::Read;

    use anyhow::Context;

    let reader = std::io::BufReader::new(file);
    let mut archive = zip::ZipArchive::new(reader)?;

    let mut header = String::new();
    archive
        .by_name("wintun/include/wintun.h")?
        .read_to_string(&mut header)?;
    header.push_str(WINTUN_BINDINGS_PREAMBLE);

    let bindings = bindgen::Builder::default()
        .header_contents("wintun.h", &header)
        .allowlist_function("Wintun.*")
        .allowlist_type("WINTUN_.*")
        .dynamic_library_name("wintun")
        .dynamic_link_require_all(true)
        .generate()
        .context("Failed to generate bindings from wintun archive")
        .unwrap();

    let mut binary = Vec::new();
    let target = std::env::var("TARGET")?;
    let arch = match target.split('-').next() {
        Some("i686") => "x86",
        Some("x86_64") => "amd64",
        Some("aarch64") => "arm64",
        Some("thumbv7a") => "arm",
        Some(a) => panic!("{} is not a supported architecture", a),
        None => unreachable!(),
    };
    archive
        .by_name(&format!("wintun/bin/{}/wintun.dll", arch))?
        .read_to_end(&mut binary)
        .context("Failed to read binary from wintun archive")?;

    Ok((bindings, binary))
}

#[cfg(windows)]
const WINTUN_FILENAME: &str = "wintun-0.14.1.zip";

#[cfg(windows)]
const WINTUN_INTEGRITY: &str = "sha256-B8JWGF1u42UuCfpVwLZz4mJLVl4CxLkJHHnKfS8k71E=";

#[cfg(windows)]
const WINTUN_BINDINGS_PREAMBLE: &str = r#"
WINTUN_CLOSE_ADAPTER_FUNC WintunCloseAdapter;
WINTUN_OPEN_ADAPTER_FUNC WintunOpenAdapter;
WINTUN_GET_ADAPTER_LUID_FUNC WintunGetAdapterLUID;
WINTUN_GET_RUNNING_DRIVER_VERSION_FUNC WintunGetRunningDriverVersion;
WINTUN_DELETE_DRIVER_FUNC WintunDeleteDriver;
WINTUN_SET_LOGGER_FUNC WintunSetLogger;
WINTUN_START_SESSION_FUNC WintunStartSession;
WINTUN_END_SESSION_FUNC WintunEndSession;
WINTUN_CREATE_ADAPTER_FUNC WintunCreateAdapter;
WINTUN_GET_READ_WAIT_EVENT_FUNC WintunGetReadWaitEvent;
WINTUN_RECEIVE_PACKET_FUNC WintunReceivePacket;
WINTUN_RELEASE_RECEIVE_PACKET_FUNC WintunReleaseReceivePacket;
WINTUN_ALLOCATE_SEND_PACKET_FUNC WintunAllocateSendPacket;
WINTUN_SEND_PACKET_FUNC WintunSendPacket;
"#;
