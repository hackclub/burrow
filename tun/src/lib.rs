#![deny(missing_debug_implementations)]

#[cfg(target_os = "windows")]
#[path = "windows/mod.rs"]
mod imp;

#[cfg(any(target_os = "linux", target_vendor = "apple"))]
#[path = "unix/mod.rs"]
pub(crate) mod imp;

mod options;
pub mod routing;

#[cfg(any(target_os = "linux", target_vendor = "apple"))]
#[cfg(feature = "tokio")]
pub mod tokio;

pub use imp::{TunInterface, TunQueue};
pub use options::TunOptions;
