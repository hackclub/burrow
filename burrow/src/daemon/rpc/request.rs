use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use tun::TunOptions;

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag="method", content="params")]
pub enum DaemonCommand {
    Start(DaemonStartOptions),
    ServerInfo,
    ServerConfig,
    Stop,
    ReloadConfig(String),
    AddConfigToml(String),
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct DaemonStartOptions {
    pub tun: TunOptions,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct DaemonRequest {
    pub id: u64,
    #[serde(flatten)]
    pub command: DaemonCommand,
}

#[test]
fn test_daemoncommand_serialization() {
    insta::assert_snapshot!(serde_json::to_string(&DaemonCommand::Start(
        DaemonStartOptions::default()
    ))
    .unwrap());
    insta::assert_snapshot!(
        serde_json::to_string(&DaemonCommand::Start(DaemonStartOptions {
            tun: TunOptions { ..TunOptions::default() }
        }))
        .unwrap()
    );
    insta::assert_snapshot!(serde_json::to_string(&DaemonCommand::ServerInfo).unwrap());
    insta::assert_snapshot!(serde_json::to_string(&DaemonCommand::Stop).unwrap());
    insta::assert_snapshot!(serde_json::to_string(&DaemonCommand::ServerConfig).unwrap())
}
