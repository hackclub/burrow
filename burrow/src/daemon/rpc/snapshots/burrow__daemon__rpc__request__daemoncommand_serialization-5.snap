---
source: burrow/src/daemon/rpc/request.rs
expression: "serde_json::to_string(&DaemonCommand::ServerConfig).unwrap()"
---
{"method":"ServerConfig"}
